### ISSUE TITLE
JSON Reporter printing NOT json

### ISSUE THREAD

**Issue by fodil-a** [dart-lang/test#1226]
Hi,

**When** repeteadly (fast) using the command `pub run test --reporter json`
**I expect**: printing JSON events, or not printing anything.
**Current behavior**: It prints the line `Waiting for current test(s) to finish.` which breaks my JSON parser, since it is not JSON.

**I believe it's from this:** https://github.com/dart-lang/test/blob/7514d98b420f15e577e52606bd85e79ffc0428cd/pkgs/test_core/lib/src/runner.dart#L230-L235

I can volunteer to fix this issue, by emitting new events ('waiting', and 'finished_waiting') when using reporter.

**Comment by fodil-a** [comment]
Fixing this bug can also fix #1216

**Comment by grouma** [comment]
This makes sense to me. Please loop in @natebosch or myself in the PR.

**Comment by fodil-a** [comment]
I have issues testing it, I tried multiple ways, but I just can't reproduce the "Waiting for ..." output in tests.  
The plan was to: 
>   1. run long test,  
>   2. send SIGTERM to the process (with kill()),  
>   3. find "Waiting for" in the process stdout.  

I tried different ways:
* with runDart,
* with pubRun,
* directly calling the Runner class, 
* using executable.

None of them could produce the "Waiting for ..." message in the process' stdout.

Do you have any pointers please ?

**Comment by natebosch** [comment]
I'm going to take a look at this and see if it's possible to print this through the reporter instead of with `print` directly.

**Comment by grouma** [comment]
One potential solution is to add a `log` method to the `Reporter` API. Then we can update the engine to use `reporter.log(Waiting for...`. @natebosch thoughts?

**Comment by fodil-a** [comment]
Hi,
first i want to thank you for trying to fix this.
If I may give my opinion, I don't think that having a `log` method in Reporter is a good thing.
The responsibility of reporting the Engine progress is owned by the Reporters.
I think that the nice way would be to completely delete this part:
https://github.com/dart-lang/test/blob/7514d98b420f15e577e52606bd85e79ffc0428cd/pkgs/test_core/lib/src/runner.dart#L226-L237

And put the above code in the engine, and instead of using "print", add a new callback, which reporters can subscribe to, the same way as the other callbacks present in Engine, just like that:
https://github.com/dart-lang/test/blob/590abf70ce2b2aed9809e0ecf64a9ba965be81a3/pkgs/test_core/lib/src/runner/engine.dart#L166-L170

The print "Waiting" is handled by each reporters, with the good format listening to the events.

For the plumbing, you will just need to subscribe to the event, like that:
https://github.com/dart-lang/test/blob/c0f4f232d4fecf5a051d3ae50e9cb141399464a6/pkgs/test_core/lib/src/runner/reporter/json.dart#L74

And to print the event, you print the way the reporter works, for example, for JsonReporter:
https://github.com/dart-lang/test/blob/c0f4f232d4fecf5a051d3ae50e9cb141399464a6/pkgs/test_core/lib/src/runner/reporter/json.dart#L278

**Comment by natebosch** [comment]
Yeah the design of the `Reporter` interface wouldn't fit well with a `log` method. I was working towards trying to log it as it if was printed in the test, but it's not super straightforward.

Adding a brand new stream on the `Engine` for runner level messages does sound interesting, I'll dig in to that some.

In the long term I would like to explore a different design for `Reporter`. It feels weird to me that each reporter is so coupled to the rest of the system, and things are duplicated in each. I'd like to see if there is a way to write it such that the `Reporter` is like a `Logger` in that you write to it. Each reporter would only be responsible for knowing how to emit the correct format, some other class would be responsible for listening on all the various streams and writing to it. This might be hard if the various reporters have differences in what they listen to and how they interpret it.
